### PR TITLE
post 14: stable structures

### PR DESCRIPTION
**Pull request by roman-kashitsyn**

### CODE REVIEW

File: posts/14-stable-structures.html.pm [hunk 1]
```diff
@@ -0,0 +1,671 @@
...
+}
+◊figure{
+◊marginnote["mn-log-interface"]{
+  The core interface of the ◊code{Cell} stable structure.
+}
+◊source-code["rust"]{
+impl<T: ◊code-ref["#storable-trait"]{Storable}, M: ◊code-ref["#memory"]{Memory}> struct ◊b{Cell}<T, M> {
+    ◊em{/// Returns the current cell value.}
+    ◊em{/// Complexity: O(1).}
+    pub fn ◊b{get}(&self, idx: usize) -> Option<Vec<u8>>;
```
Comment (dfinity-berestovskyy) [inline posts/14-stable-structures.html.pm:274]: It return `T`, right?

File: posts/14-stable-structures.html.pm [hunk 2]
```diff
@@ -0,0 +1,671 @@
...
+}
+}
+}
+◊figure[#:class "grayscale-diagram"]{
+  ◊marginnote["mn-log-figure"]{
+    A ◊code{Vec} is a growable mutable array.
+    The data representation depends on the ◊code{IS_FIXED_WIDTH} attribute of the item type.
+    If the type's representation is not fixed-width, the vector implementation must record each entry's length.
+  }
+  ◊(embed-svg "images/14-vec.svg")
```
Comment (dfinity-berestovskyy) [inline posts/14-stable-structures.html.pm:342]: In the picture, the size of the variable width item should be `max_size + size_of(len)`, i.e. `max_len + 4`

File: posts/14-stable-structures.html.pm [hunk 3]
```diff
@@ -0,0 +1,671 @@
...
+
+◊subsection-title["stable-vec"]{Stable vector}
+◊p{
+  A ◊code{Vec} is a growable mutable array similar to ◊code-ref["https://doc.rust-lang.org/std/vec/struct.Vec.html"]{std::vec::Vec}.
+  A stable vector stores its items by value, so it must know how much space it needs to allocate for each item, hence the ◊code-ref["#bounded-storable-trait"]{BoundedStorable} bound on the item type.
+}
+◊p{
+  Stable vector takes advantage of the ◊code-ref["#is-fixed-size-attribute"]{T::IS_FIXED_SIZE} attribute of the item type.
+  If the value size is not fixed, the vector allocates a few extra bytes to store the actual entry size and the ◊code-ref["#max-size-attribute"]{T::MAX_SIZE} bytes required for each value.
+  If all the values are the same size, the vector implementation uses ◊code-ref["#max-size-attribute"]{T::MAX_SIZE} for the item slot, saving up to 4 bytes per entry.
```
Comment (dfinity-berestovskyy) [inline posts/14-stable-structures.html.pm:305]:
```suggestion
  If all the values are the same size, the vector implementation uses ◊code-ref["#max-size-attribute"]{T::MAX_SIZE} for the item slot, saving 4 bytes per entry.
```

Comment (roman-kashitsyn) [reply posts/14-stable-structures.html.pm:305]: The vector uses 1 byte for the length if max_size is at most u8::MAX and 2 bytes if max_size is at most u16::MAX.

File: posts/14-stable-structures.html.pm [hunk 4]
```diff
@@ -0,0 +1,671 @@
...
+  }
+  ◊li{A ◊code-ref["#stable-log"]{StableLog} for transaction history.}
+}
+◊p{
+  The ledger metadata is relatively small; it should fit into two megabytes or sixteen WebAssembly memory pages.
+  We will allocate the metadata cell in a ◊a[#:href "#restricted-memory"]{restricted memory}, partitioning the rest of the stable memory between the accounts map and the log using the ◊a[#:href "#memory-manager"]{memory manager}.
+}
+◊p{
+  Let us put these ideas to code.
+  First, we must import a few type definitions from the library.
```
Comment (dfinity-berestovskyy) [inline posts/14-stable-structures.html.pm:532]: Seems like a trivial part, shall we just drop it?

File: posts/14-stable-structures.html.pm [hunk 5]
```diff
@@ -0,0 +1,671 @@
...
+  ◊li{
+    ◊em{Compatibility with ◊a[#:href "https://github.com/WebAssembly/multi-memory/blob/master/proposals/multi-memory/Overview.md"]{multi-memory} WebAssembly.}
+    The design should work when canisters have multiple stable memories since this feature is on the ◊a[#:href "https://forum.dfinity.org/t/proposal-wasm-native-stable-memory/15966#proposal-7"]{IC roadmap}.
+  }
+}
+◊p{
+  As a result of these goals, using the library requires planning and understanding your data shape.
+  For example, you might need to set an ◊a[#:href "#max-size-attribute"]{upper bound} on a value size that you cannot adjust without data migration.
+  Or you might have to decide how many memory pages you want to allocate for canister configuration.
+  There are other libraries with similar goals whose authors chose a slightly different set of traideoffs, such as ◊code-ref["https://crates.io/crates/ic-stable-memory"]{ic-stable-memory}.
```
Comment (dsarlis) [inline posts/14-stable-structures.html.pm:85]:
```suggestion
  There are other libraries with similar goals whose authors chose a slightly different set of trade-offs, such as ◊code-ref["https://crates.io/crates/ic-stable-memory"]{ic-stable-memory}.
```

Comment (roman-kashitsyn) [reply posts/14-stable-structures.html.pm:85]: My spell checker was happy with both variants, but it seems that trade-off is more common. Fixed.

File: posts/14-stable-structures.html.pm (hunk 1 above)
Comment (roman-kashitsyn) [reply posts/14-stable-structures.html.pm]: Great catch, fixed!

File: posts/14-stable-structures.html.pm (hunk 4 above)
Comment (roman-kashitsyn) [reply posts/14-stable-structures.html.pm:532]: One of my goals was to ensure that if you copy all the code into a file, it compiles cleanly.

Comment (dfinity-berestovskyy) [reply posts/14-stable-structures.html.pm:532]: Maybe a better approach would be to provide a link to a repo, leaving all the trivial details out. But it's fine, up to you...

File: posts/14-stable-structures.html.pm (hunk 2 above)
Comment (roman-kashitsyn) [reply posts/14-stable-structures.html.pm:361]: Fixed, thanks!

**Review by dragoljub-duric — COMMENTED** [review]
In the [images/14-btree.svg](https://github.com/roman-kashitsyn/mmapped.blog/pull/21/files#diff-67a62067a71615415bb5a3daf9e935e69853b68ee981a207571d763d33ea6c21) in the right leaf, there is a key `4`, and the root key is `10`. Should this `4` be a `14`?

**Pull request merged**